### ISSUE TITLE
Two torrent_handle(s), same id(s)

### ISSUE THREAD

**Issue by 3unjee** [arvidn/libtorrent#3200]
Greetings @arvidn,

Okay, so I'm using the `torrent_handle::id()` function in my libtorrent based application.

I thought it would be a reliable way to get a unique id per handle. But sometimes the id function seems to return the same value for two different handles. Is this a bug or a feature ?

I'm on Windows 32 bits / MinGW / gcc 4.9.

**Comment by ssiloti** [comment]
The code claims that right shifting the address of the underlying `torrent` object by 11 is safe because the size of `torrent` is greater than 1024, but I think that's off by one. 1024 is 2^10 so unless there's some subtly I'm missing it's only safe to shift 10.

**Comment by arvidn** [comment]
oops. yes, that sounds like a bug

**Comment by arvidn** [comment]
https://github.com/arvidn/libtorrent/pull/3201

**Comment by arvidn** [comment]
It's still not a fully correct solution since, on a 64 bit system, it still truncates the remaining high bits, which still leaves room for duplicate ids. at least in theory

**Comment by arvidn** [comment]
that's correct. The reliable hash is 20 bytes wide. This id is 4 bytes. It's derived from the pointer to the internal torrent object, which is unique, but it's technically 8 bytes wide. In practice pointers are 48 bits wide and with the shifting of 10 bits the underlying pointer is really 38 bits wide. This is then truncated to 32 bits.
